docs/design/federation-phase-1.md

```diff
@@ -0,0 +1,434 @@
+cases it may be complex. For example:
+
++ This workload can only be scheduled to cluster Foo. It cannot be
+   scheduled to any other clusters. (use case: sensitive workloads).
++ This workload prefers cluster Foo. But if there is no available
+   capacity on cluster Foo, it’s OK to be scheduled to cluster Bar
+   (use case: workload )
++ Seventy percent of this workload should be scheduled to cluster Foo,
+    and thirty percent should be scheduled to cluster Bar (use case:
+    vendor lock-in avoidance).  In phase one, we only introduce a
+    _clusterSelector_ field to filter acceptable clusters. In default
+    case there is no such selector and it means any cluster is
+    acceptable.
+
+Below is a sample of the YAML to create such a replication controller.
+
+``` 
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: nginx-controller
+spec:
+  replicas: 5
+  selector:
+    app: nginx
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: nginx
+        ports:
+        - containerPort: 80
+      clusterSelector: 
+      name in (Foo, Bar)
+```
+
+Currently clusterSelector (implemented as a
+[LabelSelector](../../pkg/apis/extensions/v1beta1/types.go#L704))
+only supports a simple list of acceptable clusters. Workloads will be
+evenly distributed on these acceptable clusters in phase one. After
+phase one we will define syntax to represent more advanced
+constraints, like cluster preference ordering, desired number of
+splitted workloads, desired ratio of workloads spread on different
+clusters, etc.
+
+Besides this explicit “clusterSelector” filter, a workload may have
+some implicit scheduling restrictions. For example it defines
+“nodeSelector” which can only be satisfied on some particular
+clusters. How to handle this will be addressed after phase one.
+
+## Ubernetes Services
+
+The Service API object exposed by Ubernetes is similar to service
+objects on Kubernetes. It defines the access to a group of pods. The
+Ubernetes service controller will create corresponding Kubernetes
+service objects on underlying clusters.  These are detailed in a
+separate design document: [Federated Services](federated-services.md).
+
+## Pod
+
+In phase one we only support scheduling replication controllers. Pod
+scheduling will be supported in later phase. This is primarily in
+order to keep the Ubernetes API compatible with the Kubernetes API.
+
+## ACTIVITY FLOWS
+
+## Scheduling
+
+The below diagram shows how workloads are scheduled on the Ubernetes control plane:
+
+1. A replication controller is created by the client.
+1. APIServer persists it into the storage.
+1. Cluster controller periodically polls the latest available resource
+   metrics from the underlying clusters.
+1. Scheduler is watching all pending RCs. It picks up the RC, make
+   policy-driven decisions and split it into different sub RCs.
```

Comment:
I discussed this with @quinton-hoole in person today and I think he convinced me that the approach described here is reasonable. The two key arguments were

1) The user who submits an RC (call it X) to the Ubernetes control plane may not have direct access to the underlying clusters where the pods run, but they may want to see the status of the per-cluster RCs, so we need per-cluster RC status reflected in the Ubernetes control plane somehow. We could have X's status concatenate all of the RC statuses, but this would mean X is a different kind of RC (its would have an array of statuses, one per cluster, rather than just one status), so we'd lose the symmetry with regular RCs. 

2) In theory you could build the Ubernetes control plane so that except for API objects created by users (like X in (1)) it is stateless/only uses soft state. It would work the way I described in the comment above. But it would be useful to be able to maintain some persistent state rather than being completely stateless. Two specific examples are
* the Ubernetes control plane might want to record the last time it got a status update from each cluster, so that it can detect cluster failure and move workload to a healthy cluster in response
* the Ubernetes control plane might want to record the decision made by the component this proposal calls the "Ubernetes scheduler," namely, how to divide the number of replicas requested in X amongst the various clusters, so that if the Ubernetes master fails and recovers, it can just continue executing against the scheduler decision that it was in the middle of carrying out when it failed. If you want to record this kind of state, the easiest way to do it is in objects that look exactly like ReplicationControllers and that live in the Ubernetes control plane. They represent the "intended state" that the Ubernetes control plane is driving the real clusters towards (by creating ReplicationControllers in those clusters). This also makes debugging easier, as the intended state is represented by objects rather than just in the memory of (the) controller(s) in the control plane.